### PR TITLE
feat: Capture and Restore property values during UI Update phase of Hot Reload

### PR DESCRIPTION
**Pull request by nickrandolph**
GitHub Issue (If applicable): https://github.com/unoplatform/uno/issues/13859

## PR Type

What kind of change does this PR introduce?
- Feature

## What is the current behavior?

There's no way to capture and restore property values during the ui update phase of hot reload in order to preserve state

## What is the new behavior?

An ElementMetadataUpdateHandler can register for CaptureState and RestoreState methods which can be used to extract and restore state information during the ui update phase

### <samp>🤖 Generated by Copilot at a4b9809</samp>

This pull request improves the hot reload feature for Uno Platform applications, by refactoring the code, adding new classes and methods, enhancing the UI update logic, and adding documentation and tests. It introduces the `TypeMappings` class to handle the type mapping operations, and the `TestingUpdateHandler` class to facilitate the testing of hot reload scenarios. It also adds new constructors and attributes for the update handlers, and new UI elements and test cases for the `Frame` and `TextBox` controls. It removes the obsolete `TypeMappingHelper` class and updates the references to it. It modifies the `ClientHotReloadProcessor` class to use the new classes and logic, and to improve the performance and reliability of hot reloads. It updates the `ElementUpdaterAgent` class to support new actions and attributes for the UI elements. It adds a new markdown file that documents the hot reload phases and extensibility points. It cleans up the using directives and improves the code quality and readability in several files.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI.RuntimeTests/Tests/HotReload/Frame/HRApp/Tests/Given_TextBox.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+
+	/// <summary>
+	/// Checks that when simple change to a XAML element (change Text on TextBlock) is applied to
+	/// the currently visible page, the UI update process will capture and restore property values
+	/// of specific controls (in this case the Text property of a TextBox
+	/// </summary>
+	[TestMethod]
+ 	[Ignore("Not yet working")]
+	public async Task When_Changing_TextBox()
```
Comment (jeromelaban) [inline src/Uno.UI.RuntimeTests/Tests/HotReload/Frame/HRApp/Tests/Given_TextBox.cs:26]: We'll need to ensure that restoring properties that are commonly bound is not causing issues when two-way is used

File: src/Uno.UI.RemoteControl/HotReload/ClientHotReloadProcessor.MetadataUpdate.cs [hunk 2]
```diff
@@ -206,31 +201,58 @@ private static async void ReloadWithUpdatedTypes(Type[] updatedTypes)
...
 
-								return (handler is not null || mappedType is not null) ? (fe, handler, mappedType) : default;
-							},
-						enumerateChildrenAfterMatch: true))
-				{
+							if (handler is not null)
+							{
+								if (!capturedStates.TryGetValue(key, out var dict))
+								{
+									dict = new Dictionary<string, object>();
```
Comment (jeromelaban) [inline src/Uno.UI.RemoteControl/HotReload/ClientHotReloadProcessor.MetadataUpdate.cs:227]:
```suggestion
									dict = new();
```

File: src/Uno.UI.RemoteControl/HotReload/ClientHotReloadProcessor.MetadataUpdate.cs [hunk 3]
```diff
@@ -206,31 +201,58 @@ private static async void ReloadWithUpdatedTypes(Type[] updatedTypes)
 				// This is called before the visual tree is updated
 				_ = handlerActions?.Do(h => h.Value.BeforeVisualTreeUpdate(updatedTypes)).ToArray();
 
-				// Iterate through the visual tree and either invole ElementUpdate, 
-				// or replace the element with a new one
-				foreach (
-					var (element, elementHandler, elementMappedType) in
-					EnumerateHotReloadInstances(
+				var capturedStates = new Dictionary<string, IDictionary<string, object>>();
```
Comment (jeromelaban) [inline src/Uno.UI.RemoteControl/HotReload/ClientHotReloadProcessor.MetadataUpdate.cs:204]: The value does not need to be an interface in this case.

**Review by jeromelaban — APPROVED** [review]

File: src/Uno.UI.RuntimeTests/Tests/HotReload/Frame/HRApp/Tests/Given_TextBox.cs (hunk 1 above)
Comment (nickrandolph) [reply src/Uno.UI.RuntimeTests/Tests/HotReload/Frame/HRApp/Tests/Given_TextBox.cs:26]: This is only a test scenario to validate that the extension is working.

**Comment by jeromelaban** [comment]
@Youssef1313 this test looks suspicious:

https://dev.azure.com/uno-platform/Uno%20Platform/_build/results?buildId=94299&view=ms.vss-test-web.build-test-results-tab&runId=2511100&resultId=100614&paneView=attachments

File: src/Uno.UI.RemoteControl/HotReload/ClientHotReloadProcessor.MetadataUpdate.cs (hunk 3 above)
Comment (nickrandolph) [reply src/Uno.UI.RemoteControl/HotReload/ClientHotReloadProcessor.MetadataUpdate.cs:204]: Updated

**Pull request merged**